src/main/scala/com/target/data_validator/validator/ColumnBased.scala

```diff
@@ -35,26 +35,40 @@ abstract class ColumnBased(column: String, condTest: Expression) extends CheapCh
   }
 }
 
-case class MinNumRows(minNumRows: Long) extends ColumnBased("", ValidatorBase.L0) {
+case class MinNumRows(minNumRows: Json) extends ColumnBased("", ValidatorBase.L0) {
   override def name: String = "MinNumRows"
 
-  override def substituteVariables(dict: VarSubstitution): ValidatorBase = this
+  override def substituteVariables(dict: VarSubstitution): ValidatorBase = {
+    val ret = MinNumRows(getVarSubJson(minNumRows, "minNumRows", dict))
+    getEvents.foreach(ret.addEvent)
+    ret
+  }
 
   override def configCheck(df: DataFrame): Boolean = {
-    if (minNumRows <= 0) {
-      val msg = s"MinNumRows: $minNumRows <= 0"
+
+    def notNaturalNumber(): Unit = {
+      val msg = "minNumRows must be a natural number"
       logger.error(msg)
       addEvent(ValidatorError(msg))
-      failed = true
-      true
-    } else {
-      false
     }
+
+    minNumRows.asNumber match {
+      case Some(jsonNumber) => jsonNumber.toLong match {
+        case Some(x) if x > 0 =>
+        case _ => notNaturalNumber()
+      }
+      case _ => notNaturalNumber()
+    }
+    failed
   }
 
   override def quickCheck(row: Row, count: Long, idx: Int): Boolean = {
-    failed = count < minNumRows
-    val pctError = if (failed) calculatePctError(minNumRows, count) else "0.00%"
+    // Convert to `JsonNumber` then to `Long`
+    // safe because already handled in `configCheck`
+    val minNumRowsLong = minNumRows.asNumber.get.toLong.get
```

Comment:
Probably fine, but if you have to make other changes, you _could_ cache this in a local private variable, e.g.

```scala
// in class def
var maybeMinNumRows: Option[Long] = None
// inside validation inside configCheck
maybeMinNumRows = Some(x)
// here
val minNumRowsLong = maybeMinNumRows.getOrElse { throw new RuntimeException("quickCheck executed before configCheck and the config for minNumRows was actually invalid.") }
```

Reply:
Yes, I was considering something along these lines.  This pattern repeats in quite a few places, basically for any checks that take JSON args.  I'd like to settle on a good pattern and make it consistent everywhere.

Reply:
This sounds like a good use case for ad-hoc polymorphism / type classes.

You could retain the definition of `case class MinNumRows(minNumRows: Long)` with scala primitives, and then have some meta type `VarSubstitutable[T]` defining how to read variable configurable parts out of the yaml/json parsing mechanism.

There could be a vanilla no-op `VarSubstitutable[Any]` in global scope and we could include a `VarSubstitutable[MinNumRows]` with the functionality in the `substituteVariables` method for each type we desire to override this behavior. It would just be a matter of reworking the code that does the var-sub visitor pattern into the initial parsing.

Reply:
These are good suggestions.  Let's revisit in a PR dedicated to variable substitution.